### PR TITLE
[14/n][dagster-airbyte] Implement materialization method for AirbyteCloudWorkspace

### PR DESCRIPTION
**Pull request by maximearmstrong**
## Summary & Motivation

This PR implements `AirbyteCloudWorkspace.sync_and_poll`, the materialization method for Airbyte Cloud assets. This method:
- calls `AirbyteCloudClient.sync_and_poll`
- takes the AirbyteOutput returned by `AirbyteCloudClient.sync_and_poll` and generates the asset materializations
- yields `MaterializeResult` for each expected asset and `AssetMaterialization` for each unexpected asset 
  - a connection table that was not in the connection at definitions loading time can be in the AirbyteOutput. Eg. the table was added after definitions loading time and before sync.
- logs a warning for each unmaterialized table
  - a connection table can be created at definitions loading time, but can be missing in the AirbyteOutput. Eg. the table was deleted after definitions loading time and before sync.

Can be leveraged like:

```python
from dagster_airbyte import AirbyteCloudWorkspace, airbyte_assets

import dagster as dg

airbyte_workspace = AirbyteCloudWorkspace(
    workspace_id=dg.EnvVar("AIRBYTE_CLOUD_WORKSPACE_ID"),
    client_id=dg.EnvVar("AIRBYTE_CLOUD_CLIENT_ID"),
    client_secret=dg.EnvVar("AIRBYTE_CLOUD_CLIENT_SECRET"),
)


@airbyte_assets(
    connection_id="airbyte_connection_id",
    name="airbyte_connection_name",
    group_name="airbyte_connection_name",
    workspace=airbyte_workspace,
)
def airbyte_connection_assets(context: dg.AssetExecutionContext, airbyte: AirbyteCloudWorkspace):
    yield from airbyte.sync_and_poll(context=context)


defs = dg.Definitions(
    assets=[airbyte_connection_assets],
    resources={"airbyte": airbyte_workspace},
)
```

## How I Tested These Changes

Additional tests with BK

## Changelog

[dagster-airbyte] Airbyte Cloud assets can now be materialized using the `AirbyteCloudWorkspace.sync_and_poll(…)` method in the definition of a `@airbyte_assets` decorator.

### CODE REVIEW

**Comment by maximearmstrong** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26559?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26559** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26559?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26559?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26432** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26432?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26431** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26431?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26430** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26430?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26429** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26429?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26428** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26428?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py [hunk 1]
```diff
@@ -101,14 +102,18 @@ def airbyte_connection_assets(context: dg.AssetExecutionContext, airbyte: Airbyt
...
+    dagster_airbyte_translator = dagster_airbyte_translator or DagsterAirbyteTranslator()
+
     return multi_asset(
         name=name,
         group_name=group_name,
-        can_subset=False,
+        can_subset=True,
         specs=[
-            spec
+            spec.merge_attributes(
```
Comment (maximearmstrong) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:112]: We add the translator to the metadata to reuse it in the materialization process.

Comment (dpeng817) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:112]: I know the pattern predates this PR, but I feel like it's a pretty big antipattern. Perhaps I'm missing something, but I also don't see a subprocess actually being called anywhere? I don't see why we need to attach the translator to access it, why can't the translator just be an argument that you need to pass? 

I guess the argument is that we do the same thing for dbt, likely fivetran, etc. but I think it's a pretty rough pattern that stretches our metadata concept beyond what's intended.

Comment (dpeng817) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:112]: Also, kinda weird that we don't do this in the spec loader? Feels like there's a potential footgun there if a user for whatever reason decides to use the spec loader then provide to a multi asset, won't contain the translator

**Review by dpeng817 — CHANGES_REQUESTED** [review]
I don't love the pattern but I agree that to be consistent we probably want to add the translator to the spec metadata.

I think we need to be consistent and also do it in the spec loader.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py (hunk 1 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:112]: > Also, kinda weird that we don't do this in the spec loader? Feels like there's a potential footgun there if a user for whatever reason decides to use the spec loader then provide to a multi asset, won't contain the translator

That's a good call. I agree that it should be set at the spec loader level. Currently, it is set at the asset spec level for dlt, but not for dbt and Sling - I will open a ticket about that.

> I know the pattern predates this PR, but I feel like it's a pretty big antipattern. Perhaps I'm missing something, but I also don't see a subprocess actually being called anywhere? I don't see why we need to attach the translator to access it, why can't the translator just be an argument that you need to pass?

I think the main argument here - and for all other integrations - is how XWorkspace.sync_and_poll(...) is expected to be called, which is only in the context of the asset decorator.

Passing the translator to both the decorator and sync_and_poll function is a bit awkward, and would most likely be error-prone.

```python
@airbyte_assets(
    connection_id="airbyte_connection_id",
    name="airbyte_connection_id",
    group_name="airbyte_connection_id",
    workspace=airbyte_workspace,
    dagster_airbyte_translator=CustomDagsterAirbyteTranslator(),
)
def airbyte_connection_assets(context: dg.AssetExecutionContext, airbyte: AirbyteCloudWorkspace):
    yield from airbyte.sync_and_poll(
        context=context, 
        dagster_airbyte_translator=CustomDagsterAirbyteTranslator() # potential errors here if another translator is passed
    )
```

That said, I agree that leveraging the metadata concept for that is kinda weird. I will open a ticket for that as well.

Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py]: Moved to spec loader in [40057f9](https://github.com/dagster-io/dagster/pull/26559/commits/40057f9c014b1470b18cb59cb2e1d6742c5c5a86)

Comment (cmpadden) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py]: FWIW the translator in metadata felt like anti-pattern when implementing the _dlt_ integration too, but at that point the goal was trying to be consistent with _dbt_, and as Maxime said, specifying it in both locations is not ideal.

Just want to share that I have the same sentiment.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_resources.py [hunk 2]
```diff
@@ -365,3 +371,80 @@ def test_airbyte_sync_and_poll_client_cancel_on_termination(
...
+        # Mocked AirbyteCloudClient.sync_and_poll returns API response where all connection tables are expected
+        result = materialize(
+            [my_airbyte_assets],
+            resources={"airbyte": workspace},
+        )
+        assert result.success
+        asset_materializations = [
+            event
+            for event in result.events_for_node(cleaned_connection_id)
+            if event.event_type_value == "ASSET_MATERIALIZATION"
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_resources.py:407]: nit: weird assertion. should assert against the actual dagster event type

File: python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_resources.py [hunk 3]
```diff
@@ -365,3 +371,80 @@ def test_airbyte_sync_and_poll_client_cancel_on_termination(
...
+        result = materialize(
+            [my_airbyte_assets],
+            resources={"airbyte": workspace},
+        )
+
+        assert result.success
+        asset_materializations = [
+            event
+            for event in result.events_for_node(cleaned_connection_id)
+            if event.event_type_value == "ASSET_MATERIALIZATION"
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_resources.py:427]: weird assertion here as well

File: python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/conftest.py [hunk 4]
```diff
@@ -248,3 +268,21 @@ def all_api_mocks_fixture(
...
+
+@pytest.fixture(name="airbyte_cloud_sync_and_poll")
+def sync_and_poll_fixture():
+    with patch("dagster_airbyte.resources.AirbyteCloudClient.sync_and_poll") as mocked_function:
+        # Airbyte output where all sync'd tables match the workspace data that was used to create the assets def
+        expected_airbyte_output = AirbyteOutput(
+            connection_details=SAMPLE_CONNECTION_DETAILS,
+            job_details=get_job_details_sample(status=AirbyteJobStatusType.SUCCEEDED),
+        )
+        # Airbyte output where a table is missing and an unexpected table is sync'd,
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/conftest.py:281]: sync'd -> synced

**Review by dpeng817 — COMMENTED** [review]
a few nits but overall lgtm. The broader discussion regarding translator in metadata is outside the scope of this PR.

**Review by dpeng817 — APPROVED** [review]
a few nits but overall lgtm. The broader discussion regarding translator in metadata is outside the scope of this PR.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_resources.py (hunk 2 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_resources.py]: Comments addressed in [efd66b3](https://github.com/dagster-io/dagster/pull/26559/commits/efd66b3f2782485114d714c607442261c260e59a)

**Pull request merged**